// piral-page-layouts/esm/create.js
import * as actions from './actions';
import { createElement } from 'react';
import { withApi, defaultRender, useGlobalState } from 'piral-core';
function getPageLayouts(items) {
    const layouts = {};
    if (items && typeof items === 'object') {
        Object.keys(items).forEach((name) => {
            layouts[name] = {
                pilet: undefined,
                component: items[name],
            };
        });
    }
    return layouts;
}
const DefaultWrapper = (props) => defaultRender(props.children);
function createPageWrapper(Wrapper = DefaultWrapper, fallback = 'default') {
    return (props) => {
        var _a;
        const layout = ((_a = props.meta) === null || _a === void 0 ? void 0 : _a.layout) || fallback;
        const registration = useGlobalState((s) => s.registry.pageLayouts[layout] || s.registry.pageLayouts[fallback]);
        const Layout = (registration === null || registration === void 0 ? void 0 : registration.component) || DefaultWrapper;
        return createElement(Layout, props, createElement(Wrapper, props));
    };
}
function withPageLayouts(pageLayouts, fallback) {
    return (state) => (Object.assign(Object.assign({}, state), { registry: Object.assign(Object.assign({}, state.registry), { wrappers: Object.assign(Object.assign({}, state.registry.wrappers), { page: createPageWrapper(state.registry.wrappers.page, fallback) }), pageLayouts }) }));
}
/**
 * Creates new Pilet API extensions for dealing with page layouts.
 */
export function createPageLayoutsApi(config = {}) {
    const { layouts = {}, fallback } = config;
    return (context) => {
        const findPageLayout = (name) => context.readState((s) => s.registry.pageLayouts[name]);
        context.defineActions(actions);
        context.dispatch(withPageLayouts(getPageLayouts(layouts), fallback));
        return (api) => {
            const pilet = api.meta.name;
            return {
                registerPageLayout(name, pageLayout) {
                    const current = findPageLayout(name);
                    if (!current || current.pilet === pilet) {
                        const component = withApi(context, pageLayout, api, 'pageLayout');
                        context.registerPageLayout(name, {
                            component,
                            pilet,
                        });
                    }
                    return () => api.unregisterPageLayout(name);
                },
                unregisterPageLayout(name) {
                    const current = findPageLayout(name);
                    if ((current === null || current === void 0 ? void 0 : current.pilet) === pilet) {
                        context.unregisterPageLayout(name);
                    }
                },
            };
        };
    };
}
